module/zfs/vdev_label.c

```diff
@@ -1007,6 +1007,56 @@ vdev_inuse(vdev_t *vd, uint64_t crtxg, vdev_labeltype_t reason,
 	return (state == POOL_STATE_ACTIVE);
 }
 
+static nvlist_t *
+vdev_aux_label_generate(vdev_t *vd, boolean_t reason_spare)
+{
+	nvlist_t *label;
+
+	/*
+	 * For inactive hot spares and level 2 ARC devices, we generate
+	 * a special label that identifies as a mutually shared hot
+	 * spare or l2cache device. We write the label in case of
+	 * addition or removal of hot spare or l2cache vdev (in which
+	 * case we want to revert the labels).
+	 */
+	VERIFY(nvlist_alloc(&label, NV_UNIQUE_NAME, KM_SLEEP) == 0);
+	VERIFY(nvlist_add_uint64(label, ZPOOL_CONFIG_VERSION,
+	    spa_version(vd->vdev_spa)) == 0);
+	VERIFY(nvlist_add_uint64(label, ZPOOL_CONFIG_POOL_STATE, reason_spare ?
+	    POOL_STATE_SPARE : POOL_STATE_L2CACHE) == 0);
+	VERIFY(nvlist_add_uint64(label, ZPOOL_CONFIG_GUID,
+	    vd->vdev_guid) == 0);
+
+	/*
+	 * This is merely to facilitate reporting the ashift of the
+	 * cache device through zdb. The actual retrieval of the
+	 * ashift (in vdev_alloc()) uses the nvlist
+	 * spa->spa_l2cache->sav_config (populated in
+	 * spa_ld_open_aux_vdevs()).
+	 */
+	if (!reason_spare) {
+		VERIFY(nvlist_add_uint64(label, ZPOOL_CONFIG_ASHIFT,
+		    vd->vdev_ashift) == 0);
+	}
+
+	/*
+	 * Add path information to help find it during pool import
+	 */
+	if (vd->vdev_path != NULL) {
+		VERIFY(nvlist_add_string(label, ZPOOL_CONFIG_PATH,
+		    vd->vdev_path) == 0);
+	}
+	if (vd->vdev_devid != NULL) {
+		VERIFY(nvlist_add_string(label, ZPOOL_CONFIG_DEVID,
+		    vd->vdev_devid) == 0);
+	}
+	if (vd->vdev_physpath != NULL) {
+		VERIFY(nvlist_add_string(label, ZPOOL_CONFIG_PHYS_PATH,
+		    vd->vdev_physpath) == 0);
+	}
```

Comment:
Sure this is only a code move, but while here we could switch it to fnvlist_*() to make some more compact.